pkg/util/workloadrepo/worker_test.go

```diff
@@ -132,15 +239,567 @@ func TestGlobalWorker(t *testing.T) {
+	tk := testkit.NewTestKit(t, store)
+
+	tk.MustExec("create database WORKLOAD_SCHEMA")
+	tk.MustExec("use WORKLOAD_SCHEMA")
+
+	_sessctx := wrk.getSessionWithRetry()
+	sess := _sessctx.(sessionctx.Context)
+
+	workloadTables = []repositoryTable{
+		{"INFORMATION_SCHEMA", "PROCESSLIST", samplingTable, "", "", "", ""},
+		{"INFORMATION_SCHEMA", "DATA_LOCK_WAITS", samplingTable, "", "", "", ""},
+	}
+	fillInTableNames()
+
+	now := time.Now()
+	var parts []time.Time
+
+	// This will have a partiton added for tomorrow.
+	parts = []time.Time{now.AddDate(0, 0, -1), now}
+	plTbl := getTable(t, "PROCESSLIST")
+	createTableWithParts(ctx, t, tk, plTbl, sess, parts)
+	require.True(t, validatePartitionsMatchExpected(ctx, t, sess, plTbl, parts))
+
+	// This will have partitions removed.   -3 will be removed when retention is 2, and -2 will be removed at 1.
+	parts = []time.Time{now.AddDate(0, 0, -3), now.AddDate(0, 0, -2), now.AddDate(0, 0, -1), now}
+	dlwTbl := getTable(t, "DATA_LOCK_WAITS")
+	createTableWithParts(ctx, t, tk, dlwTbl, sess, parts)
+	require.True(t, validatePartitionsMatchExpected(ctx, t, sess, dlwTbl, parts))
+
+	// start the thread
+	wrk.owner = wrk.newOwner(ownerKey, promptKey)
+	require.NoError(t, wrk.owner.CampaignOwner())
+	require.Eventually(t, func() bool {
+		return wrk.owner.IsOwner()
+	}, time.Second*10, time.Second)
+
+	// build remaining tables and start housekeeper
+	wrk.retentionDays = 2
+	fn := wrk.getHouseKeeper(ctx, getNextTick)
+	go fn()
+
+	// check paritions
+	require.Eventually(t, func() bool {
+		return validatePartitionsMatchExpected(ctx, t, sess, plTbl, []time.Time{now.AddDate(0, 0, -1), now, now.AddDate(0, 0, 1)})
+	}, time.Second*10, time.Second)
+	require.Eventually(t, func() bool {
+		return validatePartitionsMatchExpected(ctx, t, sess, dlwTbl, []time.Time{now.AddDate(0, 0, -2), now.AddDate(0, 0, -1), now, now.AddDate(0, 0, 1)})
+	}, time.Second*10, time.Second)
+
+	// Change the retention value.  We just want to see that it is still running.
+	wrk.Lock()
+	wrk.retentionDays = 1
+	wrk.Unlock()
+
+	// check partitions
+	time.Sleep(time.Second * 10)
+	require.Eventually(t, func() bool {
+		return validatePartitionsMatchExpected(ctx, t, sess, plTbl, []time.Time{now.AddDate(0, 0, -1), now, now.AddDate(0, 0, 1)})
+	}, time.Second*10, time.Second)
+	require.Eventually(t, func() bool {
+		return validatePartitionsMatchExpected(ctx, t, sess, dlwTbl, []time.Time{now.AddDate(0, 0, -1), now, now.AddDate(0, 0, 1)})
+	}, time.Second*10, time.Second)
+}
+
+func TestCalcNextTick(t *testing.T) {
+	loc := time.Now().Location()
+	require.True(t, calcNextTick(time.Date(2024, 12, 6, 23, 59, 59, 999999999, loc)) == time.Hour*2+time.Nanosecond)
+	require.True(t, calcNextTick(time.Date(2024, 12, 7, 0, 0, 0, 0, loc)) == time.Hour*2)
+	require.True(t, calcNextTick(time.Date(2024, 12, 7, 2, 0, 0, 1, loc)) == time.Hour*24-time.Nanosecond)
+	require.True(t, calcNextTick(time.Date(2024, 12, 7, 1, 59, 59, 999999999, loc)) == time.Nanosecond)
+}
```

Comment:
We better seperated these tests out. In my latest PR #58250, I need to export some internal APIs due to cyclic imports, which comes from `testkit`.

However, we can still test `calcNextTick` or other internal APIs as before, because these code did not involve `testkit`.